cirq/circuits/circuit_operation.py

```diff
@@ -0,0 +1,384 @@
+# Copyright 2020 The Cirq Developers
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""A structure for encapsulating entire circuits in an operation.
+
+A CircuitOperation is an Operation object that wraps a FrozenCircuit. When
+applied as part of a larger circuit, a CircuitOperation will execute all
+component operations in order, including any nested CircuitOperations.
+"""
+
+from typing import TYPE_CHECKING, Dict, List, Tuple
+
+import numpy as np
+
+from cirq import linalg, ops, protocols, study
+from cirq._compat import proper_repr
+
+if TYPE_CHECKING:
+    import cirq
+
+
+class CircuitOperation(ops.Operation):
+    """An operation that encapsulates a circuit.
+
+    This class captures modifications to the contained circuit, such as tags
+    and loops, to support more condensed serialization. Similar to
+    GateOperation, this type is immutable.
+    """
+
+    def __init__(self, circuit: 'cirq.FrozenCircuit') -> None:
+        """Constructs a gate that wraps a circuit.
+
+        Args:
+            circuit: The FrozenCircuit encapsulated by this operation.
+        """
+        self._circuit: 'cirq.FrozenCircuit' = circuit
+        self._repetitions: int = 1
+        self._qubit_map: Dict['cirq.Qid', 'cirq.Qid'] = {q: q for q in circuit.all_qubits()}
+        self._measurement_key_map: Dict[str, str] = {k: k for k in circuit.all_measurement_keys()}
+        self._param_resolver = study.ParamResolver(
+            {p: p for p in protocols.parameter_symbols(circuit)}
+        )
```

Comment:
This is probably possible, but the change is non-trivial since `with_qubit_mapping` etc. would need to handle implicit mappings. IIRC, this has a surprising number of edge cases to consider - I'll see if I can revive it.

Reply:
OK, implicit maps turned out less complicated than I remembered. It shifts the performance cost from `base_operation` to the `with_*` operations, but seeing as those operations call `base_operation` I think it's a net improvement.